File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/public/capabilities/runanywhere_solutions.dart
// SPDX-License-Identifier: Apache-2.0
//
// runanywhere_solutions.dart — v4 Solutions capability (T4.7 / T4.8).
//
// A "solution" is a prepackaged pipeline config — either a typed
// `SolutionConfig` proto, raw proto bytes, or YAML sugar — that the
// C++ core compiles into a GraphScheduler DAG and runs through the
// `rac_solution_*` C ABI. Mirrors the Swift / Kotlin / RN / Web
// capability shape so callers get the same API everywhere.
//
// §15 type-discipline: all `dart:ffi` work lives in
// `lib/native/dart_bridge_solutions.dart`; this capability holds an
// opaque `SolutionNativeHandle`.
//
// Usage:
//
//   final handle = await RunAnywhere.solutions.run(
//     config: SolutionConfig()..voiceAgent = VoiceAgentConfig()...,
//   );
//   handle.start();
//   handle.feed('hello');
//   handle.closeInput();
//   handle.destroy();

import 'dart:typed_data';

import 'package:runanywhere/foundation/errors/sdk_exception.dart';
import 'package:runanywhere/generated/solutions.pb.dart' as proto;
import 'package:runanywhere/native/dart_bridge.dart';
import 'package:runanywhere/native/dart_bridge_solutions.dart';
import 'package:runanywhere/native/types/basic_types.dart' show RacResultCode;

/// Lifecycle handle for a started solution.
///
/// Owns an opaque [SolutionNativeHandle] and forwards each verb to
/// the matching C ABI entry point via [DartBridgeSolutions]. Call
/// [destroy] (or any of the idempotent helpers) when finished —
/// there is no auto-finalizer in Dart FFI and dropping the reference
/// will leak the C resources.
class SolutionHandle {
  SolutionHandle._(this._handle);

  final SolutionNativeHandle _handle;

  /// True until [destroy] (or [close]) clears the underlying handle.
  bool get isAlive => _handle.isAlive;

  /// Start the underlying scheduler. Non-blocking.
  void start() => _checkRc('start', DartBridgeSolutions.start(_handle));

  /// Request a graceful shutdown. Non-blocking.
  void stop() => _checkRc('stop', DartBridgeSolutions.stop(_handle));

  /// Force-cancel the graph; returns once worker threads observe cancellation.
  void cancel() => _checkRc('cancel', DartBridgeSolutions.cancel(_handle));

  /// Signal end-of-stream on the root input edge.
  void closeInput() =>
      _checkRc('close_input', DartBridgeSolutions.closeInput(_handle));

  /// Feed one UTF-8 item into the root input edge.
  void feed(String item) {
    _requireAlive();
    final rc = DartBridgeSolutions.feed(_handle, item);
    if (rc != 0) {
      throw SDKException.invalidState(
        'rac_solution_feed failed: ${RacResultCode.getMessage(rc)}',
      );
    }
  }

  /// Cancel, join, and release native resources. Idempotent — safe to
  /// call multiple times or after [close].
  void destroy() => DartBridgeSolutions.destroy(_handle);

  /// Alias for [destroy] — gives the API a more conventional close-shape.
  void close() => destroy();

  void _requireAlive() {
    if (!_handle.isAlive) {
      throw SDKException.invalidState(
        'SolutionHandle has already been destroyed',
      );
    }
  }

  void _checkRc(String op, int rc) {
    _requireAlive();
    if (rc != 0) {
      throw SDKException.invalidState(
        'rac_solution_$op failed: ${RacResultCode.getMessage(rc)}',
      );
    }
  }
}

/// Solutions capability surface — `RunAnywhere.solutions`.
///
/// Stateless. Each `run(...)` call allocates a fresh native handle;
/// callers own the returned [SolutionHandle].
class RunAnywhereSolutions {
  RunAnywhereSolutions._();
  static final RunAnywhereSolutions _instance = RunAnywhereSolutions._();
  static RunAnywhereSolutions get shared => _instance;

  /// Construct and return a (created, not started) solution from either
  /// a typed [proto.SolutionConfig] proto or a raw [configBytes] buffer.
  /// Exactly one of [config] / [configBytes] / [yaml] must be supplied.
  ///
  /// Call [SolutionHandle.start] on the returned handle to launch worker
  /// threads. The handle owns its native resources — invoke
  /// [SolutionHandle.destroy] (or [SolutionHandle.close]) when finished.
  Future<SolutionHandle> run({
    proto.SolutionConfig? config,
    Uint8List? configBytes,
    String? yaml,
  }) async {
    _ensureReady();

    final supplied = [config, configBytes, yaml].where((v) => v != null).length;
    if (supplied != 1) {
      throw SDKException.validationFailed(
        'RunAnywhereSolutions.run requires exactly one of '
        'config / configBytes / yaml (got $supplied)',
        fieldPath: 'RunAnywhereSolutions.run.input',
      );
    }

    if (yaml != null) return _createFromYaml(yaml);

    final bytes = configBytes ?? Uint8List.fromList(config!.writeToBuffer());
    return _createFromProto(bytes);
  }

  SolutionHandle _createFromProto(Uint8List bytes) {
    if (bytes.isEmpty) {
      throw SDKException.validationFailed(
        'Solution config bytes are empty — refusing to call '
        'rac_solution_create_from_proto',
        fieldPath: 'SolutionConfig',
      );
    }

    final result = DartBridgeSolutions.createFromProto(bytes);
    if (!result.success || result.handle == null) {
      throw SDKException.invalidConfiguration(
        'rac_solution_create_from_proto failed: '
        '${RacResultCode.getMessage(result.resultCode)}',
      );
    }
    return SolutionHandle._(result.handle!);
  }

  SolutionHandle _createFromYaml(String yaml) {
    final result = DartBridgeSolutions.createFromYaml(yaml);
    if (!result.success || result.handle == null) {
      throw SDKException.invalidConfiguration(
        'rac_solution_create_from_yaml failed: '
        '${RacResultCode.getMessage(result.resultCode)}',
      );
    }
    return SolutionHandle._(result.handle!);
  }

  void _ensureReady() {
    if (!DartBridge.isInitialized) {
      throw SDKException.notInitialized();
    }
  }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/main/kotlin/com/runanywhere/sdk/public/extensions/Solutions/RunAnywhereSolutions.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Public API for L5 solutions runtime (T4.7 / T4.8).
 *
 * A "solution" is a prepackaged pipeline config — either a serialized
 * `runanywhere.v1.SolutionConfig` proto or a YAML document — that the
 * C++ core compiles into a GraphScheduler DAG and executes through the
 * `rac_solution_*` C ABI. Mirrors Swift RunAnywhere+Solutions.swift.
 *
 * Bare `RunAnywhere.runSolution(...)` top-level methods were deleted; the
 * canonical surface is `RunAnywhere.solutions.run(...)`.
 */

package com.runanywhere.sdk.public.extensions

import ai.runanywhere.proto.v1.SolutionConfig
import com.runanywhere.sdk.foundation.errors.SDKException
import com.runanywhere.sdk.native.bridge.RunAnywhereBridge
import com.runanywhere.sdk.public.RunAnywhere
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.concurrent.atomic.AtomicLong

class SolutionHandle internal constructor(
    nativeHandle: Long,
) {
    private val handleRef = AtomicLong(nativeHandle)

    val isAlive: Boolean
        get() = handleRef.get() != 0L

    suspend fun start() {
        val h = requireHandle()
        withContext(Dispatchers.IO) {
            val rc = RunAnywhereBridge.racSolutionStart(h)
            if (rc != RunAnywhereBridge.RAC_SUCCESS) {
                throw SDKException.operation("rac_solution_start failed with rc=$rc")
            }
        }
    }

    suspend fun stop() {
        val h = requireHandle()
        withContext(Dispatchers.IO) {
            val rc = RunAnywhereBridge.racSolutionStop(h)
            if (rc != RunAnywhereBridge.RAC_SUCCESS) {
                throw SDKException.operation("rac_solution_stop failed with rc=$rc")
            }
        }
    }

    suspend fun cancel() {
        val h = requireHandle()
        withContext(Dispatchers.IO) {
            val rc = RunAnywhereBridge.racSolutionCancel(h)
            if (rc != RunAnywhereBridge.RAC_SUCCESS) {
                throw SDKException.operation("rac_solution_cancel failed with rc=$rc")
            }
        }
    }

    suspend fun feed(input: String) {
        val h = requireHandle()
        withContext(Dispatchers.IO) {
            val rc = RunAnywhereBridge.racSolutionFeed(h, input)
            if (rc != RunAnywhereBridge.RAC_SUCCESS) {
                throw SDKException.operation("rac_solution_feed failed with rc=$rc")
            }
        }
    }

    suspend fun closeInput() {
        val h = requireHandle()
        withContext(Dispatchers.IO) {
            val rc = RunAnywhereBridge.racSolutionCloseInput(h)
            if (rc != RunAnywhereBridge.RAC_SUCCESS) {
                throw SDKException.operation("rac_solution_close_input failed with rc=$rc")
            }
        }
    }

    suspend fun destroy() {
        val h = handleRef.getAndSet(0L)
        if (h != 0L) {
            withContext(Dispatchers.IO) {
                RunAnywhereBridge.racSolutionDestroy(h)
            }
        }
    }

    private fun requireHandle(): Long {
        val h = handleRef.get()
        if (h == 0L) throw SDKException.invalidState("SolutionHandle has already been destroyed")
        return h
    }
}

class Solutions internal constructor() {
    suspend fun run(yaml: String): SolutionHandle {
        ensureReady()
        return withContext(Dispatchers.IO) {
            ensureNativeReady()
            val handle = RunAnywhereBridge.racSolutionCreateFromYaml(yaml)
            if (handle == 0L) {
                throw SDKException.operation("rac_solution_create_from_yaml returned a null handle")
            }
            SolutionHandle(handle)
        }
    }

    suspend fun run(configBytes: ByteArray): SolutionHandle {
        ensureReady()
        return withContext(Dispatchers.IO) {
            ensureNativeReady()
            val handle = RunAnywhereBridge.racSolutionCreateFromProto(configBytes)
            if (handle == 0L) {
                throw SDKException.operation("rac_solution_create_from_proto returned a null handle")
            }
            SolutionHandle(handle)
        }
    }

    suspend fun run(config: SolutionConfig): SolutionHandle =
        run(config.encode())

    /**
     * Mirror Swift's `Solutions.ensureReady()` —
     * `RunAnywhere+Solutions.swift:223-228`. Block the cold-start caller until
     * Phase 2 (services + HTTP) has completed so a `run(yaml=...)` issued
     * immediately after `RunAnywhere.initialize(...)` does not race against a
     * half-initialised HTTP client / model registry. Throws notInitialized if
     * Phase 1 has not run.
     */
    private suspend fun ensureReady() {
        if (!RunAnywhere.isInitialized) {
            throw SDKException.notInitialized("SDK not initialized")
        }
        RunAnywhere.ensureServicesReady()
    }

    private fun ensureNativeReady() {
        if (!RunAnywhereBridge.ensureNativeLibraryLoaded()) {
            throw SDKException.platform("Failed to load runanywhere_jni — solutions ABI unavailable")
        }
    }
}

private val SolutionsSingleton = Solutions()

val RunAnywhere.solutions: Solutions
    get() = SolutionsSingleton
